### ISSUE TITLE
Target "copy" task doesn't run in .Net Core project if project is "up-to-date"

### ISSUE THREAD

**Issue by rainersigwald** [dotnet/project-system#6301]
---
**Issue moved from microsoft/msbuild#5450**
- Please respond to @Dean-NC.

---

_From @Dean-NC on Saturday, June 20, 2020 2:10:45 PM_

### Steps to reproduce
In the attached demo project, there is an external .js file that gets linked to a destination in the project, but intentionally outside the wwwroot folder, so that it can maintain it's linked status, otherwise .Net core project system will fight and overtake it, removing the linked status after building and the copy target runs.

There's a "copy" target set to copy any changes to the original .js file to the wwwroot/js folder so the app can use it. I want this target to run regardless of whether the project is "up-to-date" or not, so I used a BeforeTargets = "BeforeBuild" (I also tried BeforeTargets = "Build", AfterTargets and all variations). **The target runs if there's changes to the project (C# or cshtml or .csproj), but changes to the linked .js file alone won't trigger the target to run.**

1. Build the project, start project without debugging, which will run the Index.cshtml page.
2. While in Visual Studio, open Linked\module1.js, and make some changes.
3. Build the project in Visual Studio. Note that the updated .js file won't get copied. If you have the Output Window verbosity set to Detailed, you will see nothing happens since the project is considered up-to-date.

Project file
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="3.1.3" />
  </ItemGroup>
  
  <ItemGroup>
    <None Include="..\module1.js" Link="Linked\module1.js" />
  </ItemGroup>
  
  <ItemGroup>
    <Content Update="wwwroot\js\common\*" Visible="False" />
    <None Update="Linked\**\*" CopyToPublishDirectory="Never" />
  </ItemGroup>
  
  <Target Name="CopyLinkedFiles" BeforeTargets="BeforeBuild">
    <Copy SourceFiles="%(None.Identity)" DestinationFolder="wwwroot\js\common" Condition="'%(None.Link)' != '' And '%(Extension)' == '.js'" SkipUnchangedFiles="true" OverwriteReadOnlyFiles="true" />
  </Target>

</Project>
```
The reason I've hidden the wwwroot\js\common files, is because these will be linked, and I don't want to accidentally change these in the common folder, but in my "Linked" folder instead.
For the ItemGroup link, I've tried "Content" instead of "None" which didn't help. 

[zzTempWeb2a.zip](https://github.com/microsoft/msbuild/files/4808008/zzTempWeb2a.zip)

### Expected behavior
In my traditional .Net framework web apps, the same kind of custom Target will always run, even if the project is up-to-date. I expect the same here.

### Actual behavior
The msbuild target doesn't run if the project is up-to-date.

### Environment data
**.NET Core SDK** (reflecting any global.json):
 **Version:   3.1.301**

Runtime Environment:
 OS Name:     Windows
 **OS Version:  10.0.18363**
 OS Platform: Windows
 RID:         **win10-x64**
 Base Path:   C:\Program Files\dotnet\sdk\3.1.301\

Host (useful for support):
  Version: 3.1.5
  Commit:  65cd789777

.NET Core SDKs installed:
  3.1.300 [C:\Program Files\dotnet\sdk]
  3.1.301 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

**Microsoft Visual Studio Community 2019
Version 16.6.0**
VisualStudio.16.Release/16.6.0+30114.105
Microsoft .NET Framework
Version 4.8.03752

Installed Version: Community

ASP.NET and Web Tools 2019   16.6.936.3669
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2020603+b3ac44798c16fff5b95dbcfe62dea84aa9a1bd72
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.6.936.3669

**Comment by ocallesp** [comment]
@Dean-NC, what is the exact target name you want to execute ? I see there are many targets CopyXXX.

If you want to react to changes to module1.js and copy it to the output directory you can use:

```
  <ItemGroup>
    <None Include="..\module1.js" Link="Linked\module1.js" CopytToOutputDirectory="PreserveNewest" />
  </ItemGroup>
```

**Comment by Dean-NC** [comment]
@ocallesp You don't want to use CopyToOutputDirectory for JavaScript files, because that copies the file to the _bin_ directory. The only time I'd use that is for an external DLL that I want added to _bin_.

I've used this concept for many years in standard ASP.Net apps, where I link a .js (and .css) file to the project, where the original file is located outside the project. When editing the linked file inside the project, changes are actually saved to the original file outside the project (the concept of "linked" files).
Then when building, this copy task copies all linked .js and/or .css files from the original location to the destination folder specified in the copy task, and as you know for .net Core web apps, this is usually the wwwroot folder (in my case it's a nested js folder with a "common" sub-folder).

The point is, the task gets run in **standard ASP.Net every time you build**, but **in Core it will only run if a C# or .cshtml or the .csproj file is changed, or it's the first time you're building**.

According the the documentation I've seen, the task should always get run, and that's where the _SkipUnchangedFiles_ attribute of the task comes in play.

**Comment by Dean-NC** [comment]
_What is the exact target name you want to execute?_

You can see above in my .csproj, that I created a custom target named _CopyLinkedFiles_, that has a BeforeTargets="BeforeBuild".

I expect this to get run before building, regardless of whether the project is considered up-to-date.

**Comment by drewnoakes** [comment]
@Dean-NC is this still an issue for you?

If you want VS to schedule a build based on files in this way, you need to advertise those files to the fast up-to-date check. There are quite a few options available to you, and they're described here:

https://github.com/dotnet/project-system/blob/main/docs/up-to-date-check.md

My initial thoughts would be to model your source JS/CSS/etc files as `UpToDateCheckInput` items, and the counterparts in the `wwwroot` folder as `UpToDateCheckOutout` items. Because these items do not contribute to the project's "main" output (i.e. the output DLL) you need to add `Set` metadata to both the input/output items. This will cause them to be considered in isolation. If the outputs are out of date with respect to the inputs, a built will be scheduled.

Are you sure you can't model these as `CopyToOutputDirectory` items, using the `TargetPath` metadata to specify they end up in the `wwwroot` folder?

**Comment by Dean-NC** [comment]
@drewnoakes thanks for your thoughts. The `UpToDateCheckInput/Output` is an interesting thought. I don't remember if I tried using `TargetPath`with `CopyToOutputDirectory`. Let me try `TargetPath` and if that doesn't work I'll try `UpToDateCheckInput/Output`. I'll let you know.

**Comment by ocallesp** [comment]
@Dean-NC can we close this issue ?

**Comment by Dean-NC** [comment]
@ocallesp Yes, you can close it. I posted my fix and details for anyone else with a similar problem.

**Comment by Dean-NC** [comment]
@drewnoakes Thanks for letting me know about those input/output project settings, I had no idea they existed.
